lucene/core/src/java/org/apache/lucene/search/LongDocValuesPointComparator.java

```diff
@@ -0,0 +1,208 @@
+ */
+
+package org.apache.lucene.search;
+
+import org.apache.lucene.document.LongPoint;
+import org.apache.lucene.index.DocValues;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.NumericDocValues;
+import org.apache.lucene.index.PointValues;
+import org.apache.lucene.util.DocIdSetBuilder;
+
+import java.io.IOException;
+import java.util.Arrays;
+
+import static org.apache.lucene.search.FieldComparator.IteratorSupplierComparator;
+
+public class LongDocValuesPointComparator extends IteratorSupplierComparator<Long> {
+    private final String field;
+    private final int numHits;
+    private final boolean reverse;
+    private final long missingValue;
+    private final long[] values;
+    private long bottom;
+    private long topValue;
+    boolean hasTopValue = false; // indicates that topValue for searchAfter is set
+    protected NumericDocValues docValues;
+    private DocIdSetIterator iterator;
+    private PointValues pointValues;
+    private HitsThresholdChecker hitsThresholdChecker = null;
+    private int maxDoc;
+    private int maxDocVisited;
+    private int updateCounter = 0;
+
+    public LongDocValuesPointComparator(String field, int numHits, boolean reverse, Long missingValue) {
+        this.field = field;
+        this.numHits = numHits;
+        this.reverse = reverse;
+        this.missingValue = missingValue != null ? missingValue : 0L;
+        this.values = new long[numHits];
+    }
+
+    @Override
+    void setHitsThresholdChecker(HitsThresholdChecker hitsThresholdChecker) {
+        this.hitsThresholdChecker = hitsThresholdChecker;
+    }
+
+    private long getValueForDoc(int doc) throws IOException {
+        if (docValues.advanceExact(doc)) {
+            return docValues.longValue();
+        } else {
+            return missingValue;
+        }
+    }
+
+    @Override
+    public int compare(int slot1, int slot2) {
+        return Long.compare(values[slot1], values[slot2]);
+    }
+
+    @Override
+    public void setTopValue(Long value) {
+        topValue = value;
+        hasTopValue = true;
+    }
+
+    @Override
+    public Long value(int slot) {
+        return Long.valueOf(values[slot]);
+    }
+
+    @Override
+    public LeafFieldComparator getLeafComparator(LeafReaderContext context) throws IOException {
+        docValues = DocValues.getNumeric(context.reader(), field);
+        iterator = docValues;
+        pointValues = context.reader().getPointValues(field);
+        maxDoc = context.reader().maxDoc();
+        maxDocVisited = 0;
+        return this;
+    }
+
+    @Override
+    public void setBottom(int slot) throws IOException {
+        this.bottom = values[slot];
+        // can't use hitsThresholdChecker.isThresholdReached() as it uses > numHits,
+        // while we want to update iterator as soon as threshold reaches numHits
+        if (hitsThresholdChecker != null && (hitsThresholdChecker.getHitsThreshold() >= numHits)) {
```

Comment:
I am not very happy about this change because of 2 reasons: 
1) We  can't use `hitsThresholdChecker.isThresholdReached` as it checks for greater than numHits, but we need to check starting with equal, as if there are no competitive docs later `setBottom` will not be called.
2) totalHitsRelation may not end up to be set to TotalHits.Relation.GREATER_THAN_OR_EQUAL_TO, as we set it only when we have  later competitive hits. 

I will keep thinking how to organize it better